packages/vite/src/node/build.ts

```diff
@@ -60,8 +67,11 @@ import { completeSystemWrapPlugin } from './plugins/completeSystemWrap'
 import { mergeConfig } from './publicUtils'
 import { webWorkerPostPlugin } from './plugins/worker'
 import { getHookHandler } from './plugins'
+import { BaseEnvironment } from './baseEnvironment'
+import type { Plugin, PluginContext } from './plugin'
+import type { RollupPluginHooks } from './typeUtils'
 
-export interface BuildOptions {
+export interface BuildEnvironmentOptions {
```

Comment:
They were separated because the `lib` option was at the top level only at first. We moved it to be per-environment as it is cleaner that way (I don't know if this should be the design for lib-mode for vite in the future, but for now it fits better to have per environment libraries compared to a toggle at the top level that is completely unrelated to the environments). We could now rename to `BuildOptions` and `DevOptions` instead of `BuildEnvironmentOptions` and `DevEnvironmentOptions` (all shared options are in `builder` during build and `server` during dev, so we may not need the word Environment in these two).

Reply:
`lib` options aside (which I feel is weird being per-environment), I figured `BuildOptions` (top-level) should be the source of truth, and then `BuildEnvironmentOptions` would be a subset of that (and extends specific properties like `createEnvironment` too). Should we structure it that way instead of the other way round where `BuildEnvironmentOptions` seems to be the source of truth?

Reply:
I think it isn't about the source of truth, but what options are per-environment and what options are global only. 
You should only be able to set per-environment options when using `BuildEnvironmentOptions`, so that is why that was the type that ends up being extended. At the top level, you can set global options and per-environment options (the defaults in the later case), so `BuildOptions` extends from `BuildEnvironmentOptions` adding the global options. This is how `resolve` options will work (I'll send a PR for this later, because `alias` and `preserveSymlinks` will be global).

But in the case of `BuildEnvironmentOptions` and `DevEnvironmentOptions`, all global options are in `builder` and `server`, so we don't need two types if that is the case.

Reply:
Ok that explanation makes sense 👍 The only thing still is that, should we remove `createEnvironment` from `BuildOptions` (the top-level options). Is there a usecase where you'd want to a set a global fallback `createEnvironment` function?

Reply:
Ah, I see what you mean. Interesting... I think this is the first option where a global default makes no sense. And now I understand your previous comment. Ok, so we actually need two different types here, and not one extending the other (or three types, with two extending a shared one). This doesn't happen to any other type though, all per-environment options can have defaults, so we can just extend the Environment type. Makes sense then to have both types. Good catch.

Reply:
Yeah I think it should be fine to have `BuildOptions = Omit<BuildEnvironmentOptions, 'createEnvironment'>` and we can change this to 3 types if we want later, or when we realize they need to be better separated.

As long as `createEnvironment` is not exposed to the top-level `build` config, i'm fine either ways.

Reply:
https://github.com/vitejs/vite/pull/16471/commits/9b847dc1441358c2ebbdfdc0a55c2da4ef734f0d

Reply:
Let's do this in main after we merge.